exchange/price_granularity_test.go

```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
 		},
 	}
 
-	price := 1.87
-	getOnePriceBucket(t, "low", low, price, "1.50")
-	getOnePriceBucket(t, "medium", medium, price, "1.80")
-	getOnePriceBucket(t, "high", high, price, "1.87")
-	getOnePriceBucket(t, "auto", auto, price, "1.85")
-	getOnePriceBucket(t, "dense", dense, price, "1.87")
-	getOnePriceBucket(t, "custom1", custom1, price, "1.86")
-
-	// test a cpm above the max in low price bucket
-	price = 5.72
-	getOnePriceBucket(t, "low", low, price, "5.00")
-	getOnePriceBucket(t, "medium", medium, price, "5.70")
-	getOnePriceBucket(t, "high", high, price, "5.72")
-	getOnePriceBucket(t, "auto", auto, price, "5.70")
-	getOnePriceBucket(t, "dense", dense, price, "5.70")
-	getOnePriceBucket(t, "custom1", custom1, price, "5.70")
-
-}
-
-func getOnePriceBucket(t *testing.T, name string, granularity openrtb_ext.PriceGranularity, price float64, expected string) {
-	t.Helper()
-	priceBucket, err := GetCpmStringValue(price, granularity)
-	if err != nil {
-		t.Errorf("Granularity: %s :: GetPriceBucketString: %s", name, err.Error())
+	// Define test cases
+	type aTest struct {
+		granularityId       string
+		inGranularity       openrtb_ext.PriceGranularity
+		expectedPriceBucket string
+		expectedError       error
+	}
+	testGroups := []struct {
+		groupDesc string
+		inCpm     float64
+		testCases []aTest
+	}{
+		{
+			"cpm below the max in every price bucket",
+			1.87,
+			[]aTest{
+				{"low", low, "1.50", nil},
+				{"medium", medium, "1.80", nil},
+				{"high", high, "1.87", nil},
+				{"auto", auto, "1.85", nil},
+				{"dense", dense, "1.87", nil},
+				{"custom1", custom1, "1.86", nil},
+			},
+		},
+		{
+			"cpm above the max in low price bucket",
+			5.72,
+			[]aTest{
+				{"low", low, "5.00", nil},
+				{"medium", medium, "5.70", nil},
+				{"high", high, "5.72", nil},
+				{"auto", auto, "5.70", nil},
+				{"dense", dense, "5.70", nil},
+				{"custom1", custom1, "5.70", nil},
+			},
+		},
+		{
+			"Precision value corner cases",
+			1.876,
+			[]aTest{
+				{
+					"Negative precision defaults to number of digits already in CPM float",
```

Comment:
I find the tests hard to reach with a long granularity id. I don't understand the setup.